### ISSUE TITLE
Jsonix in Google Apps  Script

### ISSUE THREAD

**Issue by MatthewGitter** [highsource/jsonix#200]
Hello. How can I use Jsonix in Google Apps Script (server side Google Docs automation)?

I used Jsonix to compile a REST API from its XSD files to JSON mappings. I tested OK in commandline Node.js a Javascript that creates a JS object from inline JSON, then marshals it under the Jsonix.Context for submisstion to the REST API, then unmarshals the response.

Then I created a Google Apps Script project with my script. I created another GAS project with the Jsonix.js script (as Jsonix.gs), then another GAS project with the mappings script (as API.gs). I configured my main script's project Resources with the Jsonix and API libraries' project keys. I can execute my script (verifying the library resources are used), but it fails when `Jsonix.createDocument()` is called. I inserted into the beginning of the Jsonix library script's `_jsonix_factory = function(_jsonix_xmldom, _jsonix_xmlhttprequest, _jsonix_fs)` a `Logger.log("_jsonix_xmldom: " + _jsonix_xmldom)`  line that logs `_jsonix_xmldom: undefined` , which is why `Jsonix.createDocument()` throws an error instead of returning a document.

Can I use the Google Apps Script service XmlService, or some other service in the GAS environment, instead of the _jsonix_xmldom that Jsonix expects by default? Or can I include as a library resource some functionally equivalent libraries? Or otherwise supply that functionality to Jsonix in GAS?

I expect that similar errors will result from Jsonix trying to use the `_jsonix_xmlhttprequest` and `_jsonix_fs` values that are also undefined. I see in the Jsonix.js (.gs) source code the two lines
```
// REWORK
// Node.js
```
in the `createDocument()` declaration and elsewhere. Perhaps this case I'm reporting is in active development?

Thanks for your insights.

**Comment by highsource** [comment]
Please see my answer on SO:

https://stackoverflow.com/a/51849894/303810

I have no idea about GAS and not even sure where to begin.

Is there a DOM implementation for GAS? Ho do you load modules on GAS? What is the equivalent of Node.js `require` or `<script src="..."/>` in browser?

What you basicall need to do is to pass `DOMImplementation`, `XMLSerializer`, `DOMParser` as properties of the `xmldom` dependency.

If you find out how to get these things on GAS, I can help you to figure out how to pass them to Jsonix.

**Comment by MatthewGitter** [comment]
Thanks for the clarity in your response on SO.

Google Apps Script is the scripting environment and the scripting language embedded in the Google Docs cloud platform. It's based on Javascript 1.6 with extensions cherrypicked from JS 1.7 and JS 1.8, providing a subset of the ECMAScript 5 API, executing inside Google's cloud (not in browser):
[https://en.wikipedia.org/wiki/Google_Apps_Script](https://en.wikipedia.org/wiki/Google_Apps_Script)

Instead of using (ie. in Node.js) eg.
```
var API = require ('API').API;
Jsonix = require('jsonix').Jsonix;
var context = new Jsonix.Context([API]);
var marshaller = context.createMarshaller();
```
to include and then use in a main GAS script the `jsonix` external code module and `API` mappings that I generated with Jsonix from the API XSD files, the GAS syntax is
```
var api = new API.API;
var context = new Jsonix.Jsonix.Context(api);
var marshaller = context.createMarshaller();
```
after registering the external Jsonix library (GAS script project) with the main GAS script project.

With Node.js I'd install the Jsonix module in the same filesystem as the main script, both accessible to the Node.js execution environment. In GAS I configure the main GAS project to register the external library GAS script project as a library resource for the main GAS project (setting the library project's project key in the main project's Resource > Libraries). That imports the library into the main project's namespace under the library's identifier configured in the main project.

So one way for me to provide the GAS environment, that executes the Jsonix library, with the XMLDOM functions that in Node.js would be provided by `var _xmldom = require('xmldom').xmldom;` or similar would be to create a GAS library project containing the XMLDOM Javascript files (renamed to *.gs). Then use that project's project key registered as a library resource in the Jsonix GAS project. And revise that Jsonix GAS project to refer to something like `var _xmldom = new XMLDOM.xmldom()` instead of the Jsonix factory Javascript function argv template receiving `_xmldom`. And likewise for `XMLSerializer` and `DOMParser` external libraries.

The Node.js (and Rhino, etc) XMLDOM module seems to consist of `dom-parser.js`, `dom.js`, `entities.js` and `sax.js` Javascript files which would probably be required to include in a GAS project, with the rest of the module's  files supporting the module packaging for Node.js (and Rhino, etc) but unnecessary when repackaged as a GAS library project:
https://github.com/jindw/xmldom

I don't need the `fs` module features, and it looks like the GAS environment has native alternatives to the `XmlHttpRequest` module.

**Comment by highsource** [comment]
Thank you for the clarification.

Essentially Jsonix only uses "standard DOM API", nothing fancy. The implementation is portable between browsers and Node.js `xmldom` exactly because of this.

So you don't necessarily need exactly `xmldom`, you should be able to take any DOM implementation. As long as it is more-or-less standard, Jsonix should work with it.

I've googled around for GAS and found this:

https://developers.google.com/apps-script/reference/xml-service/
https://developers.google.com/apps-script/reference/xml-service/xml-service

From the code samples it pretty much looks like standard DOM implementation. So instead of repackaging `xmldom` I would actually try to pass or wrap this `XmlService` thing and pass it to Jsonix.

Is there any way to test "Jsonix in GAS" locally?

**Comment by MatthewGitter** [comment]
If there is a way to execute GAS in a local environment it's obscure and complex. Testing Jsonix in GAS on the real Google Cloud isn't too hard, so the cost:benefit tells me to use the real thing.

Is there any great value in a GAS Jsonix library receiving references to the DOM implementation as the argument to the Jsonix factory method, rather than the Jsonix factory method (or a more local scope in Jsonix) getting a reference to the DOM implementation?

**Comment by MatthewGitter** [comment]
Jsonix as distributed receives references to its depended external libraries (XMLDOM, XmlHttpRequest, fs) in its factory function argv:
```
var _jsonix_factory = function(_jsonix_xmldom, _jsonix_xmlhttprequest, _jsonix_fs)
```

GAS is calling that function with its argv undefined. I expect it doesn't matter if I implement the alternative of providing to Jsonix a reference to those libraries via a function call in the Jsonix script body instead, right? Eg I register in my revised Jsonix library GAS script project an external "XMLDOM" library resource, then in the Jsonix.gs script:
```
var _jsonix_xmldom = XMLDOM.XMLDOM;
```

And likewise for the XmlHttpRequest and fs libraries if I decide to use their functionality in my GAS Jsonix library.

**Comment by highsource** [comment]
The reason why I've asked was this:

> Is there any great value in a GAS Jsonix library receiving references to the DOM implementation as the argument to the Jsonix factory method, rather than the Jsonix factory method (or a more local scope in Jsonix) getting a reference to the DOM implementation?

You wrote *as the argument to the Jsonix factory method* rather than *rather than the Jsonix factory method*. Basically referred to *Jsonix factory method* both times.

**Comment by highsource** [comment]
I don't quite understand the difficulty. You have to provide Jsonix with a DOM implementation. So basically you should do something like:

* Load your DOM implementation. Assuming the global variable `XMLDOM` holds it.
* Load Jsonix.
* Initialize Jsonix with `var Jsonix = _jsonix_factory(XMLDOM).Jsonix;`

Something along the lines.

`_jsonix_factory` exists due to different module loading mechanisms in different environments. Some people use `require`, some people use `define`. The patter Jsonix uses is close to UMD which attempts to be "universal".

Main Jsonix scripts simply expect either browser-provided DOM implementations of `_jsonix_xmldom` to be available.

Hope this helps.

If you want some specific changes in Jsonix, feel free to send me a pull request, I'll take a look. Or sketch in code how you'd like it to work.

**Comment by MatthewGitter** [comment]
It's the GAS environment calling `_jsonix_factory()`, and it's doing so without references to those libraries (which the env has no info about). Even if I config the Jsonix library GAS script project to use an XMLDOM library as a library resource, that's not going to provide the XMLDOM library in the `_jsonix_factory()` argv. So unless I find some way to inject the XMLDOM library reference into the GAS call to `_jsonix_factory()`, I'm going to have to revise the `_jsonix_factory()` function body to get that reference to the XMLDOM library, by explicitly assigning to the var instead of the function implicitly receiving it in its argv when the function is called by the env, ie. `var _jsonix_xmldom = XMLDOM.XMLDOM;` (after I register the XMLDOM external library).

Which should work. I was just looking to see if that approach suggested any problems to you. Seems not. Thanks.

**Comment by highsource** [comment]
You can simply invoke `var Jsonix = _jsonix_factory(XMLDOM).Jsonix;` after everything is loaded, can't you?

I don't think simply doing `var _jsonix_xmldom = XMLDOM.XMLDOM;` will work. Jsonix is "hidden" inside the `_jsonix_factory` function, `_jsonix_xmldom` is the parameter. If you set the global `_jsonix_xmldom` variable, Jsonix will still use the (`undefined`) `_jsonix_xmldom` parameter of the `_jsonix_factory` function.

**Comment by MatthewGitter** [comment]
When the GAS environment starts a script the env implicitly sets the value of the `Identifier` (whose default value is the project name of the library) that was specified in the script's Resources > Libraries to a reference to the GAS library project, so the script refers to the external library by the identifier. So I added an `XMLDOM` external GAS library project as a library resource to the `Jsonix` GAS project, then added the assignment of `_jsonix_xmldom` to the external library's identifier value:

```
var _jsonix_factory = function(_jsonix_xmldom, _jsonix_xmlhttprequest, _jsonix_fs)
{
  if(_jsonix_xmldom == undefined)
  { _jsonix_xmldom = XMLDOM;  }
```

That works, supplying the external library by the name that the rest of your Jsonix code expects. Then my script is configured to use the GAS `Jsonix` project as a library resource, and uses it eg.
```var  context = new Jsonix.Jsonix.Context([API.API])```
`API` is a GAS project the GAS script project depends on as a library resource, its GAS code consisting of the mapping JS code generated from the API's XSD files by running `jsonix-schema-compiler-full.jar` . 

To generalize your Jsonix code so it will still run in the other environments that `_jsonix_xmldom = XMLDOM;` assignment should be further wrapped in a test to detect the GAS environment, not just whenever the factory function is called without a value for `_jsonix_xmldom`. I haven't found any GAS API function dedicated to that detection, but there are various characteristics of the GAS environment that don't exist in other environments, such as `PropertiesService.getScriptProperties()` returning a `Property` object even if it's unpopulated when the script has none.

BTW I tried wrapping the GAS `XmlService` members corresponding to _jsonix_xmldom members. But the `XmlService` API is too different from eg. the standard DOM implementation for that to be a better way than just using a complete DOM implementation from a third party. Also `XmlService` doesn't include an XML serializer (marshaller), which is the core value of Jsonix to my script, though a recursive function wasn't hard to wrap around `XmlService.createElement()` for that purpose. I got an XMLDOM implementation in pure JS that's FOSS, created a GAS project, made the Jsonix GAS project depend on it as a library resource. It required only a little tweaking to account for the GAS namespace syntax.
